src/components/views/elements/crypto/VerificationQRCode.js

```diff
@@ -31,13 +34,81 @@ export default class VerificationQRCode extends React.PureComponent {
         // User verification use case only
         secret: PropTypes.string,
         otherUserKey: PropTypes.string, // Base64 key being verified
-        requestEventId: PropTypes.string,
+        otherUserDeviceKey: PropTypes.string, // Base64 key of the other user's device (optional)
+        requestEventId: PropTypes.string, // for DM verification only
     };
 
     static defaultProps = {
         action: "verify",
     };
 
+    static async getPropsForRequest(verificationRequest: VerificationRequest) {
+        const cli = MatrixClientPeg.get();
+        const myUserId = cli.getUserId();
+        const otherUserId = verificationRequest.otherUserId;
+        const myDeviceId = cli.getDeviceId();
+        const otherDevice = verificationRequest.estimatedTargetDevice;
+        const otherDeviceId = otherDevice ? otherDevice.deviceId : null;
+
+        const qrProps = {
+            secret: verificationRequest.encodedSharedSecret,
+            keyholderUserId: myUserId,
+            action: "verify",
+            keys: [], // array of pairs: keyId, base64Key
+            otherUserKey: "", // base64key
+            otherUserDeviceKey: "", // base64key
+            requestEventId: "", // we figure this out in a moment
+        };
+
+        const requestEvent = verificationRequest.requestEvent;
+        qrProps.requestEventId = requestEvent.getId()
+            ? requestEvent.getId()
+            : ToDeviceChannel.getTransactionId(requestEvent);
+
+        // Populate the keys we need depending on which direction and users are involved in the verification.
+        if (myUserId === otherUserId) {
+            if (!otherDeviceId) {
+                // New -> Existing session QR code
```

Comment:
might be clearer to phrase this in terms of which device is doing the showing/scanning - I'm not sure which way round this is from the comment

Reply:
tbh that feels more confusing, but will try.